_posts/2022-10-18-torch-linalg/torch-linalg.Rmd

```diff
@@ -0,0 +1,632 @@
+
+How do you compute linear least-squares regression? In R, using `lm()`; in `torch`, there is `linalg_lstsq()`.
+
+Where R, sometimes, hides complexity from the user, high-performance computation frameworks like `torch` tend to ask a bit more up-front effort, be it careful reading of documentation, or playing around some, or both. For example, here is the central piece of documentation for `linalg_lstsq()`, elaborating on the `driver` parameter to the function:
+
+```
+`driver` chooses the LAPACK/MAGMA function that will be used.
+For CPU inputs the valid values are 'gels', 'gelsy', 'gelsd, 'gelss'.
+For CUDA input, the only valid driver is 'gels', which assumes that A is full-rank.
+To choose the best driver on CPU consider:
+  -   If A is well-conditioned (its condition number is not too large), or you do not mind some precision loss:
+     -   For a general matrix: 'gelsy' (QR with pivoting) (default)
+     -   If A is full-rank: 'gels' (QR)
+  -   If A is not well-conditioned:
+     -   'gelsd' (tridiagonal reduction and SVD)
+     -   But if you run into memory issues: 'gelss' (full SVD).
+```
+
+
+Whether you'll need to know this will depend on the problem you're solving. But if you do, it certainly will help to have an idea what is being talked about there, if only in a high-level way.
+
+In our example problem below, we're going to be lucky. All drivers will return the same result -- but only once we'll have applied a "trick", of sorts. The book analyzes why that works; I won't do that here, to keep the post reasonably short. What we'll do instead is dig deeper into the various methods used by `linalg_lstsq()`, as well as a few others of common use.
+
+## The plan
+
+The way we'll organize this exploration is by solving a least-squares problem from scratch, making use of various matrix factorizations. Concretely, we'll approach the task:
+
+1.  By means of the so-called *normal equations*, the most direct way, in the sense that it immediately results from a mathematical statement of the problem.
+
+2.  Again, starting from the normal equations, but making use of *Cholesky factorization* in solving them.
+
+3.  Yet again, taking the normal equations for a point of departure, but proceeding by means of *LU* decomposition.
+
+4.  Fourth, employing another type of factorization -- *QR* -- that, together with the final one, accounts for the vast majority of decompositions applied "in the real world". With QR decomposition, the solution algorithm does not start from the normal equations.
+
+5.  And fifth and finally, making use of *Singular Value Decomposition* (SVD). Here, too, the normal equations are not needed.
```

Comment:
I don't think you need to say "fourth" and "fifth" given that it's a numbered list.

Reply:
Hm, not quite sure what exactly you mean, but looking at it, I see that only (4) and (5) repeat the numbering, verbally ... For consistency, I think I should remove those, really...